added capability to set profile image

dan-nyanko opened the pull request:
This accomplished by [react-native-image-picker](https://github.com/marcshilling/react-native-image-picker)

![image](https://cloud.githubusercontent.com/assets/12954045/17181576/a0ffff6c-53ee-11e6-86ac-6110694f3fd7.png)
![image](https://cloud.githubusercontent.com/assets/12954045/17181617/c308097e-53ee-11e6-8340-aa639efcdb16.png)

sophiabersani commented:
So far I do really like this. I was worried about avatar images being stored on the user objects, but at these sizes I don't think they will cause many problems.

Parse recommends the use of Parse.File to store media:
http://parseplatform.github.io/docs/js/guide/#files
If for any reason we need to send in larger data we should look into that instead.

Some other feedback:
- Due to PixelRatio, the profile images are now looking very stretched and pixelated on the ProfilePage. Images are stored in 180px, that is not the same as the 180 unit we set in the StyleSheet.
- No feedback to a user if the image fails to upload. Maybe a simple alert would do it.
- If an image fails to upload, it will be replaced in the ProfilePage, but not in the menu. The ProfilePage picture goes back to the previous one when restarting the app.

sophiabersani commented on js/router/SharedNavigator.js:
```diff
@@ -238,7 +235,7 @@ const SharedNavigator = React.createClass({
         break;
       case 'profile':
-        viewComponent = <ProfilePage {...sharedProps} />;
+        viewComponent = <ProfilePage {...sharedProps} controPanel={this._controlPanel} />;
```
This seems to be unused?

sophiabersani commented on ios/GoodQuestion/AppDelegate.m:
```diff
@@ -32,7 +32,7 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
    */
 
-  jsCodeLocation = [NSURL URLWithString:@"http://localhost:8081/index.ios.bundle?platform=ios&dev=true"];
+  jsCodeLocation = [NSURL URLWithString:@"http://192.168.1.41:8081/index.ios.bundle?platform=ios&dev=true"];
```
Left-over hardcoded local IP.

dan-nyanko commented:
@gbersani 

I have tested Parse.File and something to be aware of is that the file is stored on S3.

```
file = new Parse.File();
// some missing code that would attach image
// ...
file.save();
// attache file reference to user
user.avatar(file);
user.save();
```

Inside Parse, the file reference has properties for `name` and `url`.  The url property is the location on S3.  So one would have to a) have internet access to download the image b) setup a way to store this image to asyncStorage c) sync this image upon login/logout.  Having considered this and knowing that the user record is already synchronized and cached to local storage it makes sense to simply base64 encode the image onto the profile record.  

Thanks for catching the case on failed upload, I'm assuming you mean saving the user record?  So if Parse Server is offline, calling `user.save()` fails but the Profile Page image is updated?

sophiabersani commented:
Interesting, that's good to know on the Parse.File behaviour, definitely something we'll keep in mind when adding more media capabilities to the app.

Here are steps to reproduce the bug:
- Login
- Simulate a bad connection (airplane mode, disable wifi, etc)
- Attempt to submit a new picture
  - At this stage the app will not warn the user it failed
- Click the "back" button
- Go back to "Profile".
  - The page will now show the picture which failed to upload as their avatar.
  - The picture will also reset to the previous one after restarting the app.

dan-nyanko commented:
@gbersani - The source of react-native-image-picker shows that the image is resized to fit the dimensions specified in the API.  You can see the source code [java](https://github.com/marcshilling/react-native-image-picker/blob/master/android/src/main/java/com/imagepicker/ImagePickerModule.java#L577) and [ios](https://github.com/marcshilling/react-native-image-picker/blob/master/ios/ImagePickerManager.m#L535). The resizing of the image is why it is lower quality. I suppose that I could make the image size 50 to 150% bigger than the component dp to increase that quality but I'm not sure its worth it.  Let me know what you think.

Also, I added an alert and rolled back the image to fix the edge case you identified. Thanks!

sophiabersani commented:
@dan-nyanko The quality is lower due to PixelRatio + compression, and I believe that the avatars are still getting stretched out as well, making them look even worse.
- Pictures are being saved in a maximum of 180px h/w
- Pictures are being displayed with a minimum of 180x180

I tested it on both devices, the bigger the disparity between height/width, the lower quality the avatar will be.
eg. A 1800x900 pic will be scaled down to 180x90

You can enable cropping passing `allowsEditing` in options.
According to the docs it should work on iOS natively. It looks like it can also work on Android with some configuration. :)

sophiabersani commented:
Warning is working great for failed network requests, but it also comes up when the user cancels the "Pick Photo" modal. Is this intended?

dan-nyanko commented:
@gbersani 

The poor quality is from upscaling a small image (maxHeight 180, maxWidth 180) to the devices dp. Compression is applied each time a lossless format is saved, such as .jpeg or .png.

Like I previously stated, I don't feel its worth storing a 50% - 150% bigger image depending the the PixelRatio because a) it is persisted to storage b) it would be inconsistent across the users devices.  The PixelRatio is simply a [multiplier](https://facebook.github.io/react-native/docs/pixelratio.html#fetching-a-correctly-sized-image).

Cropping was considered but the tested library does not make it available on both platforms. I also considered a 3rd party service to perform auto-cropping, such as transloadit, but again did not feel that the application warranted such a service as the only images we are working with is a user avatar.

_If you have a method to allow cropping on Android as well, please let me know!_

I'll make a change to suppress the error on selecting 'Cancel'.

dan-nyanko committed:
adding feature to change avatar image

commit a158c0f925b62a5138d574a0b2976d03897dd612
Author: dan-nyanko <dan.nyanko@gmail.com>
Date:   Fri Aug 12 09:53:28 2016 -0400

    do not show alert upon cancel

commit 727fb8aedb6703f484b4e4ae19f7497572400ccc
Author: dan-nyanko <dan.nyanko@gmail.com>
Date:   Fri Aug 12 09:53:05 2016 -0400

    lets multiply the image by the PixelRatio; allowEditing

commit f4f7cb624248ec77e6aace752fb4404632a0d85f
Author: dan-nyanko <dan.nyanko@gmail.com>
Date:   Thu Aug 11 09:22:59 2016 -0400

    linting

commit 467c608b509b65183efd1006213f7c76a9169072
Author: dan-nyanko <dan.nyanko@gmail.com>
Date:   Wed Aug 10 17:07:03 2016 -0400

    provide warning that the avatar did not save

commit cd0142527e1834056e8b6b28e49f269890fec0c1
Author: dan-nyanko <dan.nyanko@gmail.com>
Date:   Thu Aug 4 15:16:49 2016 -0400

    remove unused prop; reset appdelegate to localhost;

commit 76f61570511ff3dd18426046fca2cf7b2631b08a
Author: dan-nyanko <dan.nyanko@gmail.com>
Date:   Wed Jul 27 11:29:41 2016 -0400

    added capability to set profile image

    commit 538f75e44684eebd5409c23c330bdefab4a69424
    Author: dan-nyanko <dan.nyanko@gmail.com>
    Date:   Wed Jul 27 10:57:15 2016 -0400

        change text

    commit 0da868e9119a61b3dac6107d809da4bd01b32625
    Author: dan-nyanko <dan.nyanko@gmail.com>
    Date:   Wed Jul 27 10:20:01 2016 -0400

        moving avatar

    commit c72186694341bfda8daf038cff97973b6a53b5e9
    Author: dan-nyanko <dan.nyanko@gmail.com>
    Date:   Wed Jul 27 10:15:06 2016 -0400

        refactor

    commit 3181102ecd8bebddf6681315e754fb218c4984ea
    Author: dan-nyanko <dan.nyanko@gmail.com>
    Date:   Tue Jul 26 17:03:00 2016 -0400

        nightly

dan-nyanko committed:
Merge branch 'master' into image-picker

dan-nyanko committed:
add missing FontAwesome.ttf in Resources xcode project

dan-nyanko committed:
re-enable ACCESS_MOCK_LOCATION;

note this needs turned off for release build
